posthog/tasks/alerts/checks.py

```diff
@@ -18,21 +19,183 @@
 from posthog.models import AlertConfiguration, Team
 from posthog.models.alert import AlertCheck
 from posthog.tasks.utils import CeleryQueue
+from posthog.schema import (
+    TrendsQuery,
+    IntervalType,
+    ChartDisplayType,
+    NodeKind,
+    AlertCalculationInterval,
+    AlertState,
+    TrendsAlertConfig,
+)
+from posthog.utils import get_from_dict_or_attr
+from posthog.caching.fetch_from_cache import InsightResult
+from posthog.clickhouse.client.limit import limit_concurrency
+from prometheus_client import Counter, Gauge
+from django.db.models import Q
+from typing import TypedDict, NotRequired
+
+
+# TODO: move the TrendResult UI type to schema.ts and use that instead
+class TrendResult(TypedDict):
+    action: dict
+    actions: list[dict]
+    count: int
+    data: list[float]
+    days: list[str]
+    dates: list[str]
+    label: str
+    labels: list[str]
+    breakdown_value: str | int | list[str]
+    aggregated_value: NotRequired[float]
+    status: str | None
+    compare_label: str | None
+    compare: bool
+    persons_urls: list[dict]
+    persons: dict
+    filter: dict
+
+
+HOURLY_ALERTS_BACKLOG_GAUGE = Gauge(
+    "hourly_alerts_backlog",
+    "Number of hourly alerts that are not being checked in the last hour.",
+)
+
+DAILY_ALERTS_BACKLOG_GAUGE = Gauge(
+    "daily_alerts_backlog",
+    "Number of daily alerts that are not being checked in the last 24 hours.",
+)
+
+API_REQUESTS_ERROR_COUNTER = Counter(
+    "alerts_check_failures",
+    "Number of alert check errors that don't notify the user",
+)
+
 
 logger = structlog.get_logger(__name__)
 
 
-def check_all_alerts() -> None:
-    # TODO: Consider aligning insight calculation with cache warming of insights, see warming.py
-    # Currently it's implicitly aligned by alerts obviously also using cache if available
+WRAPPER_NODE_KINDS = [NodeKind.DATA_TABLE_NODE, NodeKind.DATA_VISUALIZATION_NODE, NodeKind.INSIGHT_VIZ_NODE]
+
+NON_TIME_SERIES_DISPLAY_TYPES = {
+    ChartDisplayType.BOLD_NUMBER,
+    ChartDisplayType.ACTIONS_PIE,
+    ChartDisplayType.ACTIONS_BAR_VALUE,
+    ChartDisplayType.ACTIONS_TABLE,
+    ChartDisplayType.WORLD_MAP,
+}
+
+
+@shared_task(
+    ignore_result=True,
+    expires=60 * 60,
+)
+def hourly_alerts_backlog_task() -> None:
+    """
+    This runs every 5min to check for alerts with hourly check SLA
+    but haven't been checked in the last hour + 5min
+    """
+    now = datetime.now(UTC)
+
+    alerts_breaching_sla = AlertConfiguration.objects.filter(
+        Q(
+            enabled=True,
+            calculation_interval=AlertCalculationInterval.HOURLY,
+            last_checked_at__lte=now - relativedelta(hours=1, minutes=5),
+        )
+    ).count()
+
+    HOURLY_ALERTS_BACKLOG_GAUGE.set(alerts_breaching_sla)
+
+
+@shared_task(
+    ignore_result=True,
+    expires=60 * 60,
+)
+def daily_alerts_backlog_task() -> None:
```

Comment:
Agree, have combined the backlog tasks into one, though want to keep the backlog tasks separate from the alert checking logic so we can schedule them for different frequency + failures in alert checking shouldn't prevent backlog updates

Reply:
Can we also combine the alerts tasks?

Reply:
I wanted to keep them separate for now as they run at different frequencies + we don't have parallel count/limit how many alerts we fetch at one point for recalculation (will see if there's a need for it since we already stagger the alerts), so will be a bit messy to combine them

Reply:
I don't think I understand - is there any reason they need to run at difference frequencies? What stops the 2 minute job from just querying all the alerts that need to be calculated?

If we don't have a limit on the amount of alerts we fetch, isn't grabbing chunks of them at 15 minute or 1 hour intervals worse? We'd queue more up together?

Reply:
> What stops the 2 minute job from just querying all the alerts that need to be calculated?

the 'alerts that need to be calculated' are will be spread out as different ones will be eligible in that window. The next time the task runs it'll find more eligible ones.

For the hourly alert we want to check it every hour at roughly the same time. So the 2min frequency here makes sure they're queued within 2min of being eligible to recalculate + we don't want to delay their calculation more. If we also pick the daily alerts (which we would want to eventually be for longer running insights), they will get queued as well and the hourly alerts will have to keep compete with them even though we have longer to recompute daily alerts. Like if hourly alert check for last hour being 15min off will be a big deal but a daily alert check for last day being 15min off, not really.

By running the daily alert check less often more hourly alerts should complete around the same time every hour.

Sorry if it's still not clear, maybe the way I'm staggering them isn't ideal and that's what you want to improve? :)

Reply:
I think there's validity to saying that hour long alerts should have a higher priority than month long alerts, but I don't think this really makes sense the way it's constructed at the moment.

Don't all alerts go into the same celery queue after `check_alerts` runs?

So what happens when the every-2-minute hourly frequency check runs immediately after a monthly check, and the queue is completely full of long-running monthly jobs?

I think what you might actually want here is multiple queues (or some other sort of priority system), so urgent alerts get prioritized over monthly alerts?

Reply:
That all being said, I think those changes can be out of scope for the MVP - we're not trying to hit a strict SLA here yet, but the current system adds complexity and I don't think improves those guarantees.